File: LeetCode/ArrayStrings/python/hIndex.py
from typing import List


class Solution:
    def hIndex(self, citations: List[int]) -> int:
        n = len(citations)
        paper_counts = [0] * (n+1)
        
        for c in citations:
            paper_counts[min(n,c)] = paper_counts[min(n,c)] + 1

        h = n
        papers = paper_counts[n]

        while papers < h:
            h -= 1
            papers += paper_counts[h]

        return h

    @staticmethod
    def testSolution(input) -> None:
        print("Input: citations: {}".format(input["citations"]))
        print("Expected: {}".format(input["expected"]))
        print("Result: {}".format(Solution().hIndex(input["citations"])))

    
if __name__ == "__main__":
    input = {"citations": [3,0,6,1,5], "expected": 3}
    Solution.testSolution(input)
    input = {"citations": [1,3,1], "expected": 1}
    Solution.testSolution(input)

File: LeetCode/ArrayStrings/java/hIndex.java
package LeetCode.ArrayStrings.java;

import java.util.Arrays;

record hIndexRecord(int[] citations, int expected) {

}

public class hIndex {
    public int hIndexSolution(int[] citations) {
        int n = citations.length;

        int[] paper_counts = new int[n + 1];
        Arrays.fill(paper_counts, 0);

        for (int c : citations) {
            paper_counts[Math.min(n, c)] = paper_counts[Math.min(n, c)] + 1;
        }

        int h = n;
        int papers = paper_counts[n];

        while (papers < h) {
            h -= 1;
            papers += paper_counts[h];
        }

        return h;
    }

    public static void main(String[] args) {
        hIndexRecord input = new hIndexRecord(new int[] { 3, 0, 6, 1, 5 }, 3);
        hIndex.testSolution(input);
        input = new hIndexRecord(new int[] { 1, 3, 1 }, 1);
        hIndex.testSolution(input);
    }

    private static void testSolution(hIndexRecord input) {
        System.out.println("Input: citations: " + Arrays.toString(input.citations()));
        System.out.println("Expected: " + input.expected());
        System.out.println("Result: " + new hIndex().hIndexSolution(input.citations()));
        System.out.println("-".repeat(50));
    }
}
